rfcs/convergence/as-prop.md

```diff
@@ -0,0 +1,197 @@
+# RFC: Simplify handling of `as` prop
+
+---
+
+@layershifter @levithomason
+
+## Summary
+
+Restrict `as` prop to accept only HTML elements due issues with typings and problems with developer experience.
+
+## Typings problem
+
+Fluent UI components should support `as` prop to allow consumers to customize HTML tags, it's especially useful for components like `Text`:
+
+```tsx
+<>
+  {/* renders <span /> */}
+  <Text />
+  {/* renders <p /> */}
+  <Text as="p" />
+</>
+```
+
+However, currently components are accepting the whole `React.ElementType` in [`as` prop](https://github.com/microsoft/fluentui/blob/7f5086718eec496063c6830302c162117fcfc4ec/packages/react-utilities/src/compose/types.ts#L15):
+
+```tsx
+<Button as={Link} /> // renders <Link /> + additional props
+```
+
+These usages are nightmare for typings as we must use really complicated generics ([`ComponentWithAs`](https://github.com/microsoft/fluentui/blob/c27c8fa7ae589a61267bd0a6f0f1c7a49a6be9c2/packages/react-compose/src/types.ts#L21)) to support polymorphism. There are few similar implementations:
```

Comment:
```suggestion
These usages are a nightmare for typings since we must use really complicated generics in order to guarantee type safety ([`ComponentWithAs`](https://github.com/microsoft/fluentui/blob/c27c8fa7ae589a61267bd0a6f0f1c7a49a6be9c2/packages/react-compose/src/types.ts#L21)) to support polymorphism. There are few similar implementations:
```